charmstore.go

```diff
@@ -239,6 +248,148 @@ func (s *CharmStore) Latest(curls ...*charm.URL) ([]CharmRevision, error) {
 	return responses, nil
 }
 
+// ListResources returns metadata for all the resources defined on the given charms.
+func (s *CharmStore) ListResources(ids []*charm.URL) ([]ResourceResult, error) {
+	if len(ids) == 0 {
+		return nil, nil
+	}
+
+	results, err := s.client.ListResources(ids)
+	if err != nil {
+		return nil, errgo.NoteMask(err, "cannot get resource metadata from the charm store", errgo.Any)
+	}
+
+	result := make([]ResourceResult, len(ids))
+	for i, id := range ids {
+		resources, ok := results[id.String()]
+		if !ok {
+			result[i].Err = CharmNotFound(id.String())
+			continue
+		}
+		list := make([]resource.Resource, len(resources))
+		for j, res := range resources {
+			resource, err := apiResource2Resource(res)
+			if err != nil {
+				return nil, errgo.Notef(err, "got bad data from server for resource %q", res.Name)
+			}
+			list[j] = resource
+		}
+		result[i].Resources = list
+	}
+	return result, nil
+}
+
+func apiResource2Resource(res params.Resource) (resource.Resource, error) {
+	var result resource.Resource
+	resType, err := apiResourceType2ResourceType(res.Type)
+	if err != nil {
+		return result, errgo.Mask(err, errgo.Any)
+	}
+	origin, err := apiOrigin2Origin(res.Origin)
+	if err != nil {
+		return result, errgo.Mask(err, errgo.Any)
+	}
+	fp, err := resource.ParseFingerprint(res.Fingerprint)
+	if err != nil {
+		return result, errgo.Mask(err, errgo.Any)
+	}
+	return resource.Resource{
+		Meta: resource.Meta{
+			Name:        res.Name,
+			Type:        resType,
+			Path:        res.Path,
+			Description: res.Description,
+		},
+		Origin:      origin,
+		Revision:    res.Revision,
+		Fingerprint: fp,
+		Size:        res.Size,
+	}, nil
+}
+
+// UploadResource uploads the bytes from the given file as a resource with the given name for the charm.
+func (s *CharmStore) UploadResource(id *charm.URL, name, filename string) (revision int, err error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return -1, errgo.Mask(err, errgo.Any)
+	}
+	defer f.Close()
+	rev, err := s.client.UploadResource(id, name, filename, f)
+	if err != nil {
+		return rev, errgo.Mask(err, errgo.Any)
+	}
+	return rev, nil
+}
+
+// ResourceData holds the information about the bytes of a resource.
+type ResourceData struct {
```

Comment:
The one we have defined is in juju/juju ... so I don't think we can/should reference that from here.  If it were in juju/charm, then maybe, though we need to return the revision, which is not included in that other type.  Really, it's just a dumb data bag to avoid returning 4 values, and it has no functionality, so I don't think we need to worry about duplicating code here.